### PR TITLE
Improve data access for coefficients

### PR DESCRIPTION
**Pull request by IgorBaratta**
Map linear stride to unit stride in coefficients.

The idea is to try to avoid linear access (non-unit stride) within the quadrature loop.

Original code:
```c++
    for (int iq = 0; iq < 4; ++iq)
    {
        // Quadrature loop body setup for quadrature rule fad
        // Varying computations for quadrature rule fad
        double w0_d100_c0 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d100_c0 += w[ic * 3] * FE18_C0_D100_Qfad[0][0][iq][ic];
        double w0_d010_c0 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d010_c0 += w[ic * 3] * FE18_C0_D010_Qfad[0][0][iq][ic];
        double w0_d001_c0 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d001_c0 += w[ic * 3] * FE18_C0_D001_Qfad[0][0][iq][ic];
        double w0_d100_c1 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d100_c1 += w[ic * 3 + 1] * FE18_C0_D100_Qfad[0][0][iq][ic];
        double w0_d010_c1 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d010_c1 += w[ic * 3 + 1] * FE18_C0_D010_Qfad[0][0][iq][ic];
        double w0_d001_c1 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d001_c1 += w[ic * 3 + 1] * FE18_C0_D001_Qfad[0][0][iq][ic];
        double w0_d100_c2 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d100_c2 += w[ic * 3 + 2] * FE18_C0_D100_Qfad[0][0][iq][ic];
        double w0_d010_c2 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d010_c2 += w[ic * 3 + 2] * FE18_C0_D010_Qfad[0][0][iq][ic];
        double w0_d001_c2 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d001_c2 += w[ic * 3 + 2] * FE18_C0_D001_Qfad[0][0][iq][ic];ic];
```

New code:
Temporary arrays `_w_{coeff_offset}_{block_offset}` are created and accessed with unit stride inside the quadrature loop.
```c++
    double _w_0_0[10];
    for (int ic = 0; ic < 10; ++ic)
        _w_0_0[ic] = w[ic * 3];
    double _w_0_1[10];
    for (int ic = 0; ic < 10; ++ic)
        _w_0_1[ic] = w[ic * 3 + 1];
    double _w_0_2[10];
    for (int ic = 0; ic < 10; ++ic)
        _w_0_2[ic] = w[ic * 3 + 2];
    for (int iq = 0; iq < 4; ++iq)
    {
        // Quadrature loop body setup for quadrature rule fad
        // Varying computations for quadrature rule fad
        double w0_d100_c0 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d100_c0 += _w_0_0[ic] * FE18_C0_D100_Qfad[0][0][iq][ic];
        double w0_d010_c0 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d010_c0 += _w_0_0[ic] * FE18_C0_D010_Qfad[0][0][iq][ic];
        double w0_d001_c0 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d001_c0 += _w_0_0[ic] * FE18_C0_D001_Qfad[0][0][iq][ic];
        double w0_d100_c1 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d100_c1 += _w_0_1[ic] * FE18_C0_D100_Qfad[0][0][iq][ic];
        double w0_d010_c1 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d010_c1 += _w_0_1[ic] * FE18_C0_D010_Qfad[0][0][iq][ic];
        double w0_d001_c1 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d001_c1 += _w_0_1[ic] * FE18_C0_D001_Qfad[0][0][iq][ic];
        double w0_d100_c2 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d100_c2 += _w_0_2[ic] * FE18_C0_D100_Qfad[0][0][iq][ic];
        double w0_d010_c2 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d010_c2 += _w_0_2[ic] * FE18_C0_D010_Qfad[0][0][iq][ic];
        double w0_d001_c2 = 0.0;
        for (int ic = 0; ic < 10; ++ic)
            w0_d001_c2 += _w_0_2[ic] * FE18_C0_D001_Qfad[0][0][iq][ic];
 ...
}

```

### CODE REVIEW

**Review by garth-wells — APPROVED** [review]

**Review by chrisrichardson — APPROVED** [review]

File: ffcx/codegeneration/symbols.py [hunk 1]
```diff
@@ -145,11 +145,19 @@ def domain_dofs_access(self, gdim, num_scalar_dofs, restriction):
...
-    def coefficient_dof_access(self, coefficient, dof_number):
-        # TODO: Add domain number?
+    def coefficient_dof_access(self, coefficient, dof_index):
         offset = self.coefficient_offsets[coefficient]
         w = self.S("w")
-        return w[offset + dof_number]
+        return w[offset + dof_index]
+
+    def coefficient_dof_access_blocked(self, coefficient: ufl.Coefficient, index,
+                                       block_size, dof_offset, num_dofs):
```
Comment (michalhabera) [inline ffcx/codegeneration/symbols.py:154]: Unused argument `num_dofs`.

File: ffcx/codegeneration/definitions.py [hunk 2]
```diff
@@ -73,27 +73,43 @@ def coefficient(self, t, mt, tabledata, quadrature_rule, access):
...
         assert begin < end
 
         # Get access to element table
         FE = self.symbols.element_table(tabledata, self.entitytype, mt.restriction)
-
         ic = self.symbols.coefficient_dof_sum_index()
-        dof_access = self.symbols.coefficient_dof_access(mt.terminal, ic * bs + begin)
+
         code = []
+        pre_code = []
```
Comment (michalhabera) [inline ffcx/codegeneration/definitions.py:89]: Could you add a generated comment saying ~this are temporary arrays with unit stride for coefficient w_foo?

**Review by michalhabera — APPROVED** [review]

File: ffcx/codegeneration/symbols.py (hunk 1 above)
Comment (IgorBaratta) [reply ffcx/codegeneration/symbols.py]: Removed unused argument.

File: ffcx/codegeneration/definitions.py (hunk 2 above)
Comment (IgorBaratta) [reply ffcx/codegeneration/definitions.py:89]: Added a comment before the creation of the auxiliary array.

**Pull request merged**